_posts/2015-09-01-Cocoa-Architecture:-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,146 @@
+
+We have this problem with our Artwork view controller in Eigen. There are buttons that are many [stack views deep](https://speakerdeck.com/orta/ios-at-artsy?slide=38) that need to pass a message back to the view controller. When we first hit this the issue I immediately used the responder chain, you write a [line of code](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/Views/Artwork/ARArtworkActionsView.m#L85) like: `[bidButton addTarget:self action:@selector(tappedBidButton:) forControlEvents:UIControlEventTouchUpInside];` where the `self` is referring to the view. This would send the message `tappedBidButton:` up the responder chain where it is reacted upon by the [ARArtworkViewController](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/View_Controllers/Artwork/ARArtworkViewController+ButtonActions.m#L114). 
+
+I had to explain the premise of the responder chain to almost everyone touching this area of the code base. This is great in terms of the ["lucky 10,000"](https://xkcd.com/1053/) but  means that the pattern is unintuitive to those who have not previously heard of it. There was one more issue, the lack of coupling means that renaming selectors via refactoring can break the chain. 
+
+The way that we reduced the cognitive load was via a protocol, all of the actions that the responder chain will use are mapped inside [ARArtworkActionsViewButtonDelegate](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/Views/Artwork/ARArtworkActionsView.h#L10-L20)-like protocols. It's a bit of a white-lie given that there is no direct relationship using the protocol in the app, but it makes the relationship more obvious. We use a class extension that [conforms to these types of protocols](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/View_Controllers/Artwork/ARArtworkViewController+ButtonActions.h#L11) to keep the actions all kept in one place.
+
+### Wrap-up
+
+There are many design patterns, and they all come with trade-offs. Over time, our opinions on what is "good code" changes, this is fine. It's important that as developers we understand that being able to change our minds is one of the most vital skills we have in our toolchain. This means being open to opinions outside of your usual sphere of influence and to maybe bring some good ideas from them. It's great to be passionate about an aspect of how we craft applications, but try to be pragmatic over dogmatic.
```

Comment:
"this is fine" -> makes me think of http://i.ytimg.com/vi/C-zCzM5qPec/maxresdefault.jpg . Maybe say "this is good" or even "this is great", depending on how excited you are about it. Personally, I'd go for "this is great!"

"try to be pragmatic over dogmatic" -> a common dichotomy is "pragmatism vs idealism", which could be conflated in the reader's head. If you want to take a stance on that, then own it and go all the way :100: